src/chrome/utils/localizedError.ts

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import * as nls from 'vscode-nls';
+import * as path from 'path';
+import { LocalizeInfo, LocalizeFunc } from 'vscode-nls';
+import _ = require('lodash');
+import { readFileSync } from 'fs';
+import { telemetry } from '../../telemetry';
+
+// Store the key of the last call to localize, so we can inject it in the next LocalizedError we create
+const defaultLastKeyOfLocalize = "last localization key couldn't be determined";
+let lastKeyOfLocalize = defaultLastKeyOfLocalize;
+
+/** We'll replace nls.loadMessageBundle to return this custom version of localize, which stores the last key used */
+function customLocalize(file: string | undefined, localize: LocalizeFunc, info: LocalizeInfo, message: string, ...args: (string | number | boolean | undefined | null)[]): string;
+function customLocalize(file: string | undefined, localize: LocalizeFunc, key: string | number, message: string, ...args: (string | number | boolean | undefined | null)[]): string;
+function customLocalize(file: string | undefined, localize: LocalizeFunc, key: string | LocalizeInfo | number, message: string, ...args: (string | number | boolean | undefined | null)[]): string {
+    lastKeyOfLocalize = extractLocalizationKey(key, file);
+
+    return localize(<string>key, message, ...args);
+}
+
+/** Format of the bundle.json file */
+interface LocalizationBundle {
+    [file: string]: { keys: (string | undefined)[] } | undefined;
+}
+
+/** We store the bundle metadata on memory to be able to extract the localization keys based on the file and message index */
+let bundle: LocalizationBundle = {};
+const bundlePath = path.resolve(__dirname, '../../../nls.metadata.json');
```

Comment:
This is not safe, the shape of this file and the transformations that vscode-nls/vscode-nls-dev do are not part of their API, they are internal implementation details. They have changed in the past and can change again.

Reply:
That's a good point. Changing to a different approach so we don't rely on the format of that file.